src/binary_heap.rs

```diff
@@ -1086,6 +1097,47 @@ impl<T, C: Compare<T>> BinaryHeap<T, C> {
 
         self.rebuild_tail(start);
     }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all elements `e` for which `f(&e)` returns
+    /// `false`. The elements are visited in unsorted (and unspecified) order.
+    ///
+    /// # Examples
+    ///
+    /// Basic usage:
+    ///
+    /// ```
+    /// use std::collections::BinaryHeap;
+    ///
+    /// let mut heap = BinaryHeap::from([-10, -5, 1, 2, 4, 13]);
+    ///
+    /// heap.retain(|x| x % 2 == 0); // only keep even numbers
+    ///
+    /// assert_eq!(heap.into_sorted_vec(), [-10, 2, 4])
+    /// ```
+    // #[stable(feature = "binary_heap_retain", since = "1.70.0")]
+    pub fn retain<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&T) -> bool,
+    {
+        // rebuild_start will be updated to the first touched element below, and the rebuild will
+        // only be done for the tail.
+        let mut guard = RebuildOnDrop {
+            rebuild_from: self.len(),
+            heap: self,
+        };
+        let mut i = 0;
+
+        guard.heap.data.retain(|e| {
+            let keep = f(e);
+            if !keep && i < guard.rebuild_from {
+                guard.rebuild_from = i;
+            }
+            i += 1;
+            keep
+        });
```

Comment:
FWIW, I'm not fond of stdlib's usage of separate `Drop` impls, since they make the reasoning non-local. Would you mind using https://docs.rs/unwind-safe instead (we already have that dep in our dep tree).

```suggestion
        ::unwind_safe::with_state((self, self.len())
            .try_eval(|(this, rebuild_from)| {
                // rebuild_from will be updated to the first touched element below,
                // and the rebuild will only be done for the tail.
                let mut i = 0;
                this.data.retain(|e| {
                    let keep = f(e);
                    if !keep && i < *rebuild_from {
                        *rebuild_from = i;
                    }
                    i += 1;
                    keep
                });
            })
            .finally(|(this, rebuild_from)| {
                this.rebuild_tail(rebuild_from);
            })
```

But obviously feel free to disregard this suggestion if the objective is just to stick to being a stdlib quasi-copy 🙂

Reply:
I think that would prevent us from submitting a PR upstream.

Reply:
I guess, we have to make decision on whether we consider submitting upstream. The upstream doesn't seem to be active, and with Arthur latest suggestion, we might only need the `remove_once` instead of porting `retain` over.